### DISCUSSION TITLE
Hardware: Do we need anti-aliasing filters?

### DISCUSSION THREAD

**Discussion by corrados** [corrados/edrumulus#6] [Ideas]
In the current Edrumulus hardware design, there is no anti-aliasing filter considered. I measured the spectrum of a typical piezo trigger signal and it did not have a lot of high frequencies. It looks like most of the signal is located below 4 kHz. 

However, electrical interference signals might be present which might have higher frequency components. The signal interferer situation may be improved with an anti-aliasing filter.

For a possible anti-aliasing filter, the way the samples are captured in the micro controller should be considered. E.g., for the Teensy hardware, 8 samples are averaged in the ADC hardware. So, we sample at 8 kHz but at every sample point, the signal is sampled very quickly 8 times which results in a non-even sampling with a high sampling rate. I am not sure what influence this averaging has on the design of a possible analog anti-aliasing filter.

**Comment by jstma** [comment]
I couldn't find any details on the adc hardware in the i.MX RT1062.  The [datasheet](https://www.nxp.com/docs/en/nxp/data-sheets/IMXRT1060CEC.pdf) doesn't say much about the actual adc hardware.  It does have a handy picture:

![Screenshot at 2021-10-11 10-28-05](https://user-images.githubusercontent.com/87273793/136832805-5442c421-99b8-4e41-bb62-424e47ba8427.png)

The switched capacitor acts like a resistor with a max value of 1k according to the datasheet.

Otherwise, the datasheet really doesn't contain much in the way of details.  Oversampling and then decimating/filter will act as antialiasing, but I don't understand what you're saying when the signal is sampled quickly 8 times.  Is that just the successive approximation of this adc or something else?  Is there some documentation I can reference?

**Reply by jstma** [reply]
And before I forget, these are the Radin values (kohms):
```
                    typical    max
ADLPC=0, ADHSC=1       5        7
ADLPC=0, ADHSC=0      12.5     15
ADLPC=1, ADHSC=0      25       30
```

**Comment by jstma** [comment]
I did some more research and found this [appnote](https://www.silabs.com/documents/public/application-notes/an118.pdf) which describes something called oversampling and averaging.  Sounds a lot like what the Teensy hardware might be capable of.

The theory says that adc resolution can be improved:

"Assuming the noise approximates white noise, the random variable representing the noise is equally distributed with zero mean between ADC codes."

To increase the resolution of measurements by oversampling:
```
fos = 4^w * fs
```
where w is the number of additional bits.  They then propose that
```
SNR=6.02*ENOB + 1.76
```
This formula appears in the teeny cpu's datasheet too where they say the ENOB is 10.1 (minimum) when AVGE=1, AVGS=11.  I don't know what AVGS=11 means.  So I don't know if we can take advantage of increased resolution.

However, that app note also says:
```
In a practical sense, we measure a signal band-lim-
ited to less than 1/2*fs, then oversample that signal
with an oversampling ratio (OSR). The resulting
samples are then averaged (or decimated) for the
resulting output data.
```

Which means they require an antialiasing filter and the techinque doesn't do any antialiasing on it's own.  White noise above nyquist will reflect below nyquist even with an antialiasing filter and this oversample and average technique will reduce that aliased white noise.

I think adding an antialias filter would be a good idea.

**Reply by corrados** [reply]
Yes, I agree with you. It would be good to experiment with analog anti-aliasing filters to see if that improves the trigger detection in any way. At least theoretically, it seems to be useful.

**Reply by jstma** [reply]
Well the only thing I see it doing is preventing spurious noise from aliasing into your piezo readings.  But at the same time the oversampling should allow a simple single pole filter to work much more effectively.  Maybe.  Averaging is not quite the same as decimating.

**Comment by corrados** [comment]
I just took the "Teensy noise signal" I recorded, cut it in half (so that I only see the Teensy noise), normalized the signal in Audacity and plotted the spectrum:
![grafik](https://user-images.githubusercontent.com/46655886/143762974-bcd4d5be-b01f-461f-8fcd-8341118a0225.png)
As you can see, at high frequencies, the noise spectrum goes up a lot. That is strange. Does anybody have an explanation of this effect?

**Reply by jstma** [reply]
It looks like aliasing,  If the noise is on the adc power pin as I suspect, an antialiasing filter won't help.  The hardware on the Teensy 4.0 needs to be fixed.  However, if the noise is coming into the adc pin, then an antialias filter might help.

I've been meaning to pick up a Teensy 3.6 to see if the adc noise is better, but they were out of stock last time I looked.
